pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/SegmentPreprocessThrottler.java

```diff
@@ -0,0 +1,212 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.local.utils;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import java.util.Map;
+import java.util.Set;
+import org.apache.pinot.common.concurrency.AdjustableSemaphore;
+import org.apache.pinot.spi.config.provider.PinotClusterConfigChangeListener;
+import org.apache.pinot.spi.utils.CommonConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Used to throttle the total concurrent index rebuilds that can happen on a given Pinot server.
+ * Code paths that do no need to rebuild the index or which don't happen on the server need not utilize this throttler.
+ */
+public class SegmentPreprocessThrottler implements PinotClusterConfigChangeListener {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentPreprocessThrottler.class);
+
+  /**
+   * _maxPreprocessConcurrency and _maxConcurrentPreprocessesBeforeServingQueries must be >= 0. To effectively disable
+   * throttling, this can be set to a very high value
+   */
+  private int _maxPreprocessConcurrency;
+  private int _maxPreprocessConcurrencyBeforeServingQueries;
+  private boolean _relaxThrottling;
+  private final AdjustableSemaphore _semaphore;
+
+  /**
+   * @param maxPreprocessConcurrency configured preprocessing concurrency
+   * @param maxPreprocessConcurrencyBeforeServingQueries configured preprocessing concurrency before serving queries
+   * @param relaxThrottling whether to relax throttling prior to serving queries
+   */
+  public SegmentPreprocessThrottler(int maxPreprocessConcurrency, int maxPreprocessConcurrencyBeforeServingQueries,
+      boolean relaxThrottling) {
```

Comment:
Consider naming it `isServingQueries` to be more specific

Reply:
done